src/MongoDB.Driver.Core/Core/NativeLibraryLoader/NativeLibrary.cs

```diff
@@ -0,0 +1,166 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using MongoDB.Driver.Core.Misc;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace MongoDB.Driver.Core.NativeLibraryLoader
+{
+    internal enum SupportedPlatforms
+    {
+        Windows,
+        Linux,
+        MacOS
+    }
+
+    internal interface INativeLibrary
+    {
+        T GetFunction<T>(string name);
+    }
+
+    internal class NativeLibrary : INativeLibrary
+    {
+        private readonly string _libraryName;
+        private readonly INativeLibraryLoader _loader;
+
+        public NativeLibrary(IDictionary<SupportedPlatforms, string> libraryRelativePaths, string libraryName)
```

Comment:
Instead of passing in these parameters let's pass in a:

    Func<SupportedPlatform, string> libraryLocator

function that this class will call when it needs to locate the shared library file for the current platform.

See my prototype branch.